### PR TITLE
Prevent use of uninitialized scalar Parameters in JIT code (#7847, partial)

### PR DESCRIPTION
**Pull request by steven-johnson**

### CODE REVIEW

File: test/correctness/func_clone.cpp [hunk 1]
```diff
@@ -159,14 +159,6 @@ int update_defined_after_clone_test() {
         return 1;
     }
 
-    Buffer<int> im = g.realize({200, 200});
```
Comment (abadams) [inline test/correctness/func_clone.cpp:162]: Prefix this block of code with param.set(false) instead of deleting it.

Comment (steven-johnson) [reply test/correctness/func_clone.cpp:162]: Done

File: src/Parameter.cpp [hunk 2]
```diff
@@ -20,6 +20,7 @@ struct ParameterContents {
     std::vector<BufferConstraint> buffer_constraints;
     Expr scalar_default, scalar_min, scalar_max, scalar_estimate;
     const bool is_buffer;
+    bool data_ever_set = false;
```
Comment (TH3CHARLie) [inline src/Parameter.cpp:23]: this is the place we realize whenever we add sth to the frontend IR, serialization needs to be updated at the same time

Comment (steven-johnson) [reply src/Parameter.cpp:23]: I don't understand

Comment (TH3CHARLie) [reply src/Parameter.cpp:23]: `data_ever_set` looks like a field that needs to be consistent across serialization (i.e. A parameter that has a data set should also be known as data-ever-set after serialization roundtrip). In this case, we need to add this field to the serialization implementation to properly serialize it.

Comment (steven-johnson) [reply src/Parameter.cpp:23]: Hmm... didn't think about that. 

I'm actually kinda unsure as to whether we should be saving the scalar value of the Param (if any) to the Serialization or not -- I know we do now (and it's required for the serialization-round-trip-via-JIT hack), but *conceptually*, it seems like the wrong thing to do. 

A Parameter (either scalar or buffer) is conceptually just a placeholder with required-type information. Is it really desirable (in the general sense) to save whatever happens to be stuff in there?

Comment (TH3CHARLie) [reply src/Parameter.cpp:23]: it does seem weird to me at first as well, had to admit it.

But now I treat this new field in the PR same/similar to the purpose of `defined`, then it makes more sense

**Comment by steven-johnson** [comment]
PTAL -- this change ballooned quite a bit but I think this is all better now:
- Use `std::optional<halide_scalar_value_t>` instead of uint64-and-bool for the scalar data field.
- Mark the Flatbuffer field as `= null` so that the API for it uses `std::optional` as well. (Note that there is no binary format change here since all Flatbuffer fields are optional by default.)
- Make the recently-added-for-deserialization ctors for Parameter private, with friend access for Deserializer; also remove the redundant `is_buffer` arguments since each of them has only one legal value anyway (and delete the Python wrappers for them entirely)
- Tighten the implementation of `Parameter:scalar_expr()` and `Parameter::scalar()`

PTAL

File: src/Pipeline.cpp [hunk 3]
```diff
@@ -808,6 +825,11 @@ void Pipeline::prepare_jit_call_arguments(RealizationArg &outputs, const Target
                                           bool is_bounds_inference, JITCallArgs &args_result) {
     user_assert(defined()) << "Can't realize an undefined Pipeline\n";
 
+    FindVariablesWithUnsetParams find_unset;
```
Comment (abadams) [inline src/Pipeline.cpp:828]: I think this code is a hot path, so we can't do IR traversal here.

File: src/Pipeline.cpp [hunk 4]
```diff
@@ -70,6 +70,23 @@ std::string sanitize_function_name(const std::string &s) {
     return name;
 }
 
+class FindVariablesWithUnsetParams : public IRVisitor {
+    using IRVisitor::visit;
+
+    void visit(const Variable *op) override {
+        if (op->param.defined() && !op->param.is_buffer() && op->param.name() != "__user_context") {
+            user_assert(op->param.has_scalar_value())
```
Comment (abadams) [inline src/Pipeline.cpp:78]: I don't think this error message is right. Won't it trigger on all attempts to realize a pipeline with unset params, regardless of whether or not you're in a generator? I think the much more likely scenario is someone using the JIT forgetting to set the value of a Param<T>. Also, you can absolutely compile_to_callable without setting the scalar params.

File: src/Pipeline.cpp (hunk 3 above)
Comment (steven-johnson) [reply src/Pipeline.cpp:828]: Blarg. That's no good. I'm open for better suggestions.

File: src/Pipeline.cpp (hunk 4 above)
Comment (steven-johnson) [reply src/Pipeline.cpp:78]: > Won't it trigger on all attempts to realize a pipeline with unset params, regardless of whether or not you're in a generator?

Yes. That's a feature, not a bug.

> I don't think this error message is right.

Yeah, this is a bit confusing -- IIRC the reason for wording it this way is that the "other" cases (realize, callable) end up being caught via `read_only_scalar_address()` noting the invalid value and failing (apparently they don't add the interposed Variable, I don't recall why, it's been a few days).

**Comment by steven-johnson** [comment]
OK, so upon further review:

`FindVariablesWithUnsetParams` was added to try to address the case of using `compile_to_callable()` inside a Generator, but I think it's basically impossible to (mis)use `compile_to_callable()` in this way inside a Generator; you'd have to construct your own `Argument` directly to use in `compile_to_callable()` (since Generator Input fields don't expose any conversion to Argument or Param<T> or Parameter)... and even then, there isn't a way to invoke the Callable with the Input, because the Callable requires a `float` argument (an `Expr` won't do). So I don't think we even need that traverser (or the uninitialized_param_3 test) at all.

**Review by abadams — APPROVED** [review]

**Pull request merged**